UB in shared lock acqusition (BucketSnapshotManager)

Caught by thread safety static analysis, this is on latest `master`

```
	mv -f $depbase.Tpo $depbase.Po
bucket/BucketSnapshotManager.cpp:150:24: error: cannot call function 'copySearchableLiveBucketListSnapshot' while mutex 'mSnapshotMutex' is held [-Werror,-Wthread-safety-analysis]
        liveSnapshot = copySearchableLiveBucketListSnapshot();
                       ^
bucket/BucketSnapshotManager.cpp:155:30: error: cannot call function 'copySearchableHotArchiveBucketListSnapshot' while mutex 'mSnapshotMutex' is held [-Werror,-Wthread-safety-analysis]
        hotArchiveSnapshot = copySearchableHotArchiveBucketListSnapshot();
                             ^
2 errors generated.
make[3]: *** [bucket/BucketSnapshotManager.o] Error 1
make[3]: *** Waiting for unfinished jobs....
make[2]: *** [all] Error 2
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
```

--- a/.github/workflows/clang-19-static-analysis.yml
+++ b/.github/workflows/clang-19-static-analysis.yml
@@ -1,0 +1,103 @@
+name: Clang-19 Static Analysis
+
+on:
+  pull_request:
+  merge_group:
+  push:
+    branches:
+      - master
+      - prod
+      - testnet
+      - acceptance-test-pass
+
+jobs:
+  clang-19-compile:
+    runs-on: ubuntu-latest
+    env:
+      CACHED_PATHS: |
+        ~/.cargo
+        target
+    strategy:
+      fail-fast: false
+      matrix:
+        protocol: ["current", "next"]
+    steps:
+      - name: Fix kernel mmap rnd bits
+        # Asan in llvm provided in ubuntu 22.04 is incompatible with
+        # high-entropy ASLR in much newer kernels that GitHub runners are
+        # using leading to random crashes: https://reviews.llvm.org/D148280
+        run: sudo sysctl vm.mmap_rnd_bits=28
+
+      - name: Probe Cache
+        id: cache
+        uses: actions/cache/restore@v4
+        with:
+          path: ${{ env.CACHED_PATHS }}
+          key: ${{ runner.os }}-clang-19-${{ matrix.protocol }}-${{ github.sha }}
+          lookup-only: true
+
+      - name: Restore Cache
+        if: steps.cache.outputs.cache-hit != 'true'
+        uses: actions/cache/restore@v4
+        with:
+          path: ${{ env.CACHED_PATHS }}
+          key: ${{ steps.cache.outputs.cache-primary-key }}
+          restore-keys: |
+            ${{ runner.os }}-clang-19-${{ matrix.protocol }}
+
+      - uses: actions/checkout@v4
+        if: steps.cache.outputs.cache-hit != 'true'
+        with:
+          fetch-depth: 200
+          submodules: true
+
+      - name: install core packages
+        if: steps.cache.outputs.cache-hit != 'true'
+        run: |
+          sudo apt-get update
+          sudo apt-get -y install --no-install-recommends apt-utils dialog git iproute2 procps lsb-release
+
+      - name: install clang-19
+        if: steps.cache.outputs.cache-hit != 'true'
+        run: |
+          sudo apt-get update
+          sudo apt-get -y install clang-19 llvm-19
+
+      - name: install dependencies
+        if: steps.cache.outputs.cache-hit != 'true'
+        run: |
+          sudo apt-get -y install libc++-19-dev libc++abi-19-dev
+          sudo apt-get -y install postgresql git build-essential pkg-config autoconf automake libtool bison flex libpq-dev parallel sed perl ccache
+
+      - name: install rustup components
+        if: steps.cache.outputs.cache-hit != 'true'
+        run: rustup component add rustfmt
+
+      - name: install cargo-cache
+        if: steps.cache.outputs.cache-hit != 'true'
+        run: cargo install --locked cargo-cache --version 0.8.3
+
+      - name: install cargo-sweep
+        if: steps.cache.outputs.cache-hit != 'true'
+        run: cargo install --locked cargo-sweep --version 0.7.0
+
+      - name: Build with ci-build.sh
+        if: steps.cache.outputs.cache-hit != 'true'
+        run: |
+          # Set compiler environment variables for ci-build.sh
+          export CC='clang'
+          export CXX='clang++'
+          export CLANG_VERSION='19'
+
+          # Skip clang-format check since we don't have clang-format-12
+          export SKIP_FORMAT_CHECK=1
+
+          # Run ci-build.sh with tests disabled and protocol specified
+          ./ci-build.sh --disable-tests --protocol ${{ matrix.protocol }}
+
+      - uses: actions/cache/save@v4
+        if: ${{ steps.cache.outputs.cache-hit != 'true' && github.ref_name == 'master' }}
+        with:
+          path: ${{ env.CACHED_PATHS }}
+          key: ${{ steps.cache.outputs.cache-primary-key }}
+

--- a/ci-build.sh
+++ b/ci-build.sh
@@ -77,12 +77,14 @@ hash -r
 
 if test $CXX = 'clang++'; then
     RUN_PARTITIONS=$(seq 0 $((NPROCS-1)))
-    which clang-12
-    ln -s `which clang-12` bin/clang
-    which clang++-12
-    ln -s `which clang++-12` bin/clang++
-    which llvm-symbolizer-12
-    ln -s `which llvm-symbolizer-12` bin/llvm-symbolizer
+    # Use CLANG_VERSION environment variable if set, otherwise default to 12
+    CLANG_VER=${CLANG_VERSION:-12}
+    which clang-${CLANG_VER}
+    ln -s `which clang-${CLANG_VER}` bin/clang
+    which clang++-${CLANG_VER}
+    ln -s `which clang++-${CLANG_VER}` bin/clang++
+    which llvm-symbolizer-${CLANG_VER}
+    ln -s `which llvm-symbolizer-${CLANG_VER}` bin/llvm-symbolizer
     clang -v
     llvm-symbolizer --version || true
 elif test $CXX = 'g++'; then
@@ -97,7 +99,7 @@ fi
 
 config_flags="--enable-asan --enable-extrachecks --enable-ccache --enable-sdfprefs --enable-threadsafety ${PROTOCOL_CONFIG}"
 export CFLAGS="-O2 -g1 -fno-omit-frame-pointer -fsanitize-address-use-after-scope -fno-common"
-export CXXFLAGS="-w $CFLAGS"
+export CXXFLAGS="$CFLAGS"
 
 # quarantine_size_mb / malloc_context_size : reduce memory usage to avoid
 # crashing in tests that churn a lot of memory
@@ -130,13 +132,15 @@ ccache -s
 date
 time ./autogen.sh
 time ./configure $config_flags
-make format
-d=`git diff | wc -l`
-if [ $d -ne 0 ]
-then
-    echo "clang format must be run as part of the pull request, current diff:"
-    git diff
-    exit 1
+if [ -z "${SKIP_FORMAT_CHECK}" ]; then
+    make format
+    d=`git diff | wc -l`
+    if [ $d -ne 0 ]
+    then
+        echo "clang format must be run as part of the pull request, current diff:"
+        git diff
+        exit 1
+    fi
 fi
 
 crlf=$(find . ! \( -type d -o -path './.git/*' -o -path './Builds/*' -o -path './lib/*' \) -print0 | xargs -0 -n1 -P9 file "{}" | grep CRLF || true)

--- a/src/bucket/BucketSnapshotManager.cpp
+++ b/src/bucket/BucketSnapshotManager.cpp
@@ -147,12 +147,12 @@ BucketSnapshotManager::maybeCopyLiveAndHotArchiveSnapshots(
     SharedLockShared guard(mSnapshotMutex);
     if (needsUpdate(liveSnapshot, mCurrLiveSnapshot))
     {
-        liveSnapshot = copySearchableLiveBucketListSnapshot();
+        liveSnapshot = copySearchableLiveBucketListSnapshot(guard);
     }
 
     if (needsUpdate(hotArchiveSnapshot, mCurrHotArchiveSnapshot))
     {
-        hotArchiveSnapshot = copySearchableHotArchiveBucketListSnapshot();
+        hotArchiveSnapshot = copySearchableHotArchiveBucketListSnapshot(guard);
     }
 }
 

--- a/src/overlay/FlowControl.h
+++ b/src/overlay/FlowControl.h
@@ -133,13 +133,13 @@ class FlowControl
 
 #ifdef BUILD_TESTS
     FlowControlCapacity&
-    getCapacity()
+    getCapacity() NO_THREAD_SAFETY_ANALYSIS
     {
         return mFlowControlCapacity;
     }
 
     FlowControlCapacity&
-    getCapacityBytes()
+    getCapacityBytes() NO_THREAD_SAFETY_ANALYSIS
     {
         return mFlowControlBytesCapacity;
     }
@@ -152,7 +152,7 @@ class FlowControl
     }
 
     FloodQueues<QueuedOutboundMessage>&
-    getQueuesForTesting()
+    getQueuesForTesting() NO_THREAD_SAFETY_ANALYSIS
     {
         return mOutboundQueues;
     }

